File: TeamCode/src/main/java/org/firstinspires/ftc/common/hardware/subsystems/SubsystemManager.java
package org.firstinspires.ftc.common.hardware.subsystems;

import com.qualcomm.robotcore.hardware.HardwareMap;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class SubsystemManager {
    private static final SubsystemManager INSTANCE = new SubsystemManager();
    private List<Subsystem> subsystems;

    private SubsystemManager() {
        subsystems = new ArrayList<>();
    }

    public static SubsystemManager getInstance() {
        return INSTANCE;
    }

    public void add(Subsystem... subsystem) {
        subsystems.addAll(Arrays.asList(subsystem));
    }

    public void clear() {
        subsystems = new ArrayList<>();
    }

    public void init(HardwareMap hwMap) {
        for (Subsystem subsystem : subsystems) {
            subsystem.init(hwMap);
        }
    }

    public void run() {
        int timestamp = (int) (System.nanoTime() * 1E-6);

        for (Subsystem subsystem : subsystems) {
            if (timestamp - subsystem.getLastReadTimestamp() >= subsystem.getReadIntervalMs()) {
                subsystem.read();
                subsystem.setLastReadTimestamp(timestamp);
            }
            if (timestamp - subsystem.getLastComputeTimestamp() >= subsystem.getComputeIntervalMs()) {
                subsystem.compute();
                subsystem.setLastComputeTimestamp(timestamp);
            }
            if (timestamp - subsystem.getLastWriteTimestamp() >= subsystem.getWriteIntervalMs()) {
                subsystem.write();
                subsystem.setLastWriteTimestamp(timestamp);
            }
        }
    }
}

File: TeamCode/src/main/kotlin/brickbot/quickstart/subsystems/SubsystemManager.kt
package brickbot.quickstart.subsystems

import com.qualcomm.robotcore.hardware.HardwareMap

object SubsystemManager {
    private val subsystems = mutableMapOf<String, Subsystem>()

    /**
     * This method adds the subsystems to the subsystem map.
     * The subsystem manager does not allow multiple instances of the same subsystem class.
     * @param subsystems The subsystems to add to the map.
     */
    fun add(vararg subsystems: Subsystem) {
        for (subsystem in subsystems) {
            this.subsystems[subsystem::class.java.simpleName] = subsystem
        }
    }

    fun clear() {
        subsystems.clear()
    }

    fun init(hwMap: HardwareMap) {
        subsystems.values.forEach { it.init(hwMap) }
    }

    fun run() {
        val timestamp: Int = (System.nanoTime() * 1e-6).toInt()

        subsystems.values.forEach { it.update(timestamp) }
    }
}